From cb3bf66aadb25742a161be8c8abcb598ea27d8f4 Mon Sep 17 00:00:00 2001
From: Dinh Nguyen <dinguyen@altera.com>
Date: Tue, 26 Feb 2013 14:08:00 -0600
Subject: [PATCH 203/254] arm: socfpga: Add cpu1-start-addr to device tree

Upstream: git://git.rocketboards.org/linux-socfpga.git
 implementation

Convert the usage of CPU1 start address to a device tree implementation.
Also, clean up the unnecessary defines for multiple boards in the
altera_dt_match table.
"altr, socfpga" is all we need in the socfpga.c file.

Add documentation for the Altera's reset and system manager bindings.

Signed-off-by: Dinh Nguyen <dinguyen@altera.com>
(cherry picked from commit 7bb66e6a0cabf95a46abd02dcd329f17d4e4b8a8)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 .../bindings/arm/altera/socfpga-system.txt         |    2 +
 arch/arm/boot/dts/socfpga_cyclone5.dts             |    6 ++++-
 arch/arm/boot/dts/socfpga_ice.dts                  |    6 ++++-
 arch/arm/boot/dts/socfpga_vt.dts                   |    6 ++++-
 arch/arm/mach-socfpga/platsmp.c                    |   18 ++++++++-------
 arch/arm/mach-socfpga/socfpga.c                    |   23 ++++---------------
 6 files changed, 32 insertions(+), 29 deletions(-)

diff --git a/Documentation/devicetree/bindings/arm/altera/socfpga-system.txt b/Documentation/devicetree/bindings/arm/altera/socfpga-system.txt
index 07c65e3..f4d04a0 100644
--- a/Documentation/devicetree/bindings/arm/altera/socfpga-system.txt
+++ b/Documentation/devicetree/bindings/arm/altera/socfpga-system.txt
@@ -3,9 +3,11 @@ Altera SOCFPGA System Manager
 Required properties:
 - compatible : "altr,sys-mgr"
 - reg : Should contain 1 register ranges(address and length)
+- cpu1-start-addr : CPU1 start address in hex.
 
 Example:
 	 sysmgr@ffd08000 {
 		compatible = "altr,sys-mgr";
 		reg = <0xffd08000 0x1000>;
+		cpu1-start-addr = <0xffd080c4>;
 	};
diff --git a/arch/arm/boot/dts/socfpga_cyclone5.dts b/arch/arm/boot/dts/socfpga_cyclone5.dts
index d29addc..f93049e 100644
--- a/arch/arm/boot/dts/socfpga_cyclone5.dts
+++ b/arch/arm/boot/dts/socfpga_cyclone5.dts
@@ -20,7 +20,7 @@
 
 / {
 	model = "Altera SOCFPGA Cyclone V";
-	compatible = "altr,socfpga-cyclone5";
+	compatible = "altr,socfpga-cyclone5", "altr,socfpga";
 
 	chosen {
 		bootargs = "console=ttyS0,57600";
@@ -104,6 +104,10 @@
 				};
 			};
 
+		sysmgr@ffd08000 {
+			cpu1-start-addr = <0xffd080c4>;
+		};
+
 		timer0@ffc08000 {
 			clock-frequency = <100000000>;
 		};
diff --git a/arch/arm/boot/dts/socfpga_ice.dts b/arch/arm/boot/dts/socfpga_ice.dts
index b7b9573..7a60687 100644
--- a/arch/arm/boot/dts/socfpga_ice.dts
+++ b/arch/arm/boot/dts/socfpga_ice.dts
@@ -20,7 +20,7 @@
 
 / {
 	model = "Altera SOCFPGA Emulator";
-	compatible = "altr,socfpga-ice";
+	compatible = "altr,socfpga-ice", "altr,socfpga";
 
 	chosen {
 		bootargs = "console=ttyS0,4800";
@@ -94,6 +94,10 @@
 				};
 			};
 
+		sysmgr@ffd08000 {
+			cpu1-start-addr = <0xffd080c4>;
+		};
+
 		timer0@ffc08000 {
 			clock-frequency = <77161>;
 		};
diff --git a/arch/arm/boot/dts/socfpga_vt.dts b/arch/arm/boot/dts/socfpga_vt.dts
index acd1e14..ecd0e45 100644
--- a/arch/arm/boot/dts/socfpga_vt.dts
+++ b/arch/arm/boot/dts/socfpga_vt.dts
@@ -20,7 +20,7 @@
 
 / {
 	model = "Altera SOCFPGA VT";
-	compatible = "altr,socfpga-vt";
+	compatible = "altr,socfpga-vt", "altr,socfpga";
 
 	chosen {
 		bootargs = "console=ttyS0,57600";
@@ -99,6 +99,10 @@
 				};
 			};
 
+		sysmgr@ffd08000 {
+			cpu1-start-addr = <0xffd08010>;
+		};
+
 		timer0@ffc08000 {
 			clock-frequency = <7000000>;
 		};
diff --git a/arch/arm/mach-socfpga/platsmp.c b/arch/arm/mach-socfpga/platsmp.c
index cee352c..3ff7e89 100644
--- a/arch/arm/mach-socfpga/platsmp.c
+++ b/arch/arm/mach-socfpga/platsmp.c
@@ -44,17 +44,19 @@ static int __cpuinit socfpga_boot_secondary(unsigned int cpu, struct task_struct
 {
 	int trampoline_size = &secondary_trampoline_end - &secondary_trampoline;
 
-	memcpy(phys_to_virt(0), &secondary_trampoline, trampoline_size);
+	if (cpu1start_addr) {
+		memcpy(phys_to_virt(0), &secondary_trampoline, trampoline_size);
 
-	__raw_writel(virt_to_phys(v7_secondary_startup),
-		(sys_manager_base_addr+(cpu1start_addr & 0x000000ff)));
+		__raw_writel(virt_to_phys(v7_secondary_startup),
+			(sys_manager_base_addr+(cpu1start_addr & 0x000000ff)));
 
-	flush_cache_all();
-	smp_wmb();
-	outer_clean_range(0, trampoline_size);
+		flush_cache_all();
+		smp_wmb();
+		outer_clean_range(0, trampoline_size);
 
-	/* This will release CPU #1 out of reset.*/
-	__raw_writel(0, rst_manager_base_addr + 0x10);
+		/* This will release CPU #1 out of reset.*/
+		__raw_writel(0, rst_manager_base_addr + 0x10);
+	}
 
 	return 0;
 }
diff --git a/arch/arm/mach-socfpga/socfpga.c b/arch/arm/mach-socfpga/socfpga.c
index db4ed5b..1150089 100644
--- a/arch/arm/mach-socfpga/socfpga.c
+++ b/arch/arm/mach-socfpga/socfpga.c
@@ -92,16 +92,6 @@ static void __init socfpga_scu_map_io(void)
 	iotable_init(&scu_io_desc, 1);
 }
 
-static void __init init_socfpga_vt(void)
-{
-	cpu1start_addr = 0xffd08010;
-}
-
-static void __init init_socfpga(void)
-{
-	cpu1start_addr = 0xffd080c4;
-}
-
 static void __init enable_periphs(void)
 {
 	/* Release all peripherals from reset.*/
@@ -226,6 +216,11 @@ static void __init socfpga_sysmgr_init(void)
 	struct device_node *np;
 
 	np = of_find_compatible_node(NULL, NULL, "altr,sys-mgr");
+
+	if (of_property_read_u32(np, "cpu1-start-addr",
+			(u32 *) &cpu1start_addr))
+		pr_err("SMP: Need cpu1-start-addr in device tree.\n");
+
 	sys_manager_base_addr = of_iomap(np, 0);
 
 	np = of_find_compatible_node(NULL, NULL, "altr,rst-mgr");
@@ -260,11 +255,6 @@ static void __init gic_init_irq(void)
 	of_irq_init(irq_match);
 	socfpga_sysmgr_init();
 
-	if (of_machine_is_compatible("altr,socfpga-vt"))
-		init_socfpga_vt();
-	else
-		init_socfpga();
-
 	socfpga_init_clocks();
 	twd_local_timer_of_register();
 }
@@ -298,9 +288,6 @@ static void __init socfpga_cyclone5_init(void)
 
 static const char *altera_dt_match[] = {
 	"altr,socfpga",
-	"altr,socfpga-cyclone5",
-	"altr,socfpga-vt",
-	"altr,socfpga-ice",
 	NULL
 };
 
-- 
1.7.5.4

